[BUG]: listresolver tls does not work

### Bug Description / What Happened

`listresolver` directives using `tls://` do not work, resulting in the Client using the next available resolver. If CoreDNS is on a server that uses itself as the resolver, then no lists will be resolved.

### Expected Behavior

`listresolver` directives using `tls://` to actually use TLS and resolve lists

### Steps to reproduce

1. Use regular DNS

```nginx
. {
    filter {
        listresolver 9.9.9.9
        block list domain https://example.com/list
    }
}
```

With the above Corefile, the list is resolved using Quad9, downloaded, then loaded.

2.

```nginx
. {
    filter {
        listresolver tls://9.9.9.9
        block list domain https://example.com/list
    }
}
```

With the above Corefile, TLS connections fail resulting in the HTTP client trying the next available resolver, usually the system.

### Additional Info

A solution has been worked out and a pull request will follow shortly.

### OS

Linux

### Version

0.2.2

### Logs

```txt
Oct 02 19:12:12 ns2 systemd[1]: Started CoreDNS.
Oct 02 19:12:13 ns2 coredns[29448]: [ERROR] plugin/filter: there was a problem fetching block domain list "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt"; error fetching list "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt"; Get "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt": dial tcp: lookup s3.amazonaws.com on 127.0.0.1:53: EOF
Oct 02 19:12:13 ns2 coredns[29448]: [ERROR] plugin/filter: there was a problem fetching block domain list "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt"; error fetching list "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt"; Get "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt": dial tcp: lookup s3.amazonaws.com on 127.0.0.1:53: EOF
Oct 02 19:12:13 ns2 coredns[29448]: [ERROR] plugin/filter: there was a problem fetching block domain list "https://raw.githubusercontent.com/ookangzheng/dbl-oisd-nl/master/dbl.txt"; error fetching list "https://raw.githubusercontent.com/ookangzheng/dbl-oisd-nl/master/dbl.txt"; Get "https://raw.githubusercontent.com/ookangzheng/dbl-oisd-nl/master/dbl.txt": dial tcp: lookup raw.githubusercontent.com on 127.0.0.1:53: EOF
Oct 02 19:12:14 ns2 coredns[29448]: [ERROR] plugin/filter: there was a problem fetching block hosts list "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts"; error fetching list "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts"; Get "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts": dial tcp: lookup raw.githubusercontent.com on 127.0.0.1:53: EOF
Oct 02 19:12:14 ns2 coredns[29448]: [ERROR] plugin/filter: there was a problem fetching block hosts list "http://sysctl.org/cameleon/hosts"; error fetching list "http://sysctl.org/cameleon/hosts"; Get "http://sysctl.org/cameleon/hosts": dial tcp: lookup sysctl.org on 127.0.0.1:53: EOF
Oct 02 19:12:14 ns2 coredns[29448]: [INFO] plugin/filter: Successfully updated filter; 1 allowed domains, 0 allowed regular expressions; 0 blocked domains, 0 blocked regular expressions
Oct 02 19:12:14 ns2 coredns[29448]: .:53
Oct 02 19:12:14 ns2 coredns[29448]: CoreDNS-1.10.0
Oct 02 19:12:14 ns2 coredns[29448]: linux/arm, go1.19.1, 596a9f9-dirty
Oct 02 19:12:22 ns2 coredns[29448]: [INFO] SIGTERM: Shutting down servers then terminating
Oct 02 19:12:22 ns2 systemd[1]: Stopping CoreDNS...
Oct 02 19:12:22 ns2 systemd[1]: coredns.service: Succeeded.
Oct 02 19:12:22 ns2 systemd[1]: Stopped CoreDNS.
```

--- a/README.md
+++ b/README.md
@@ -74,7 +74,7 @@ filter {
 
 ```nginx
 filter {
-    listresolver [ RESOLVER ]
+    listresolver RESOLVER [ SERVER_NAME ]
 }
 ```
 
@@ -82,6 +82,9 @@ filter {
 and `tls` schemes are accepted. Ports may be specified. IPv6 address are
 accepted when used with a scheme and port. Since `listresolver` is intended to
 be used when no other resolvers are available, only IP addresses are accepted.
+* **SERVER_NAME**: Only used when resolver scheme is `tls`. Must be the host
+name of the resolver, otherwise resolving will fail due to being unable to
+verify the resolver's certificate.
 
 ## Domain Matching
 
@@ -125,7 +128,7 @@ solutions. Zone and Unbound configuration files are not supported.
 # hours.
 
 filter {
-    listresolver tls://9.9.9.9
+    listresolver tls://9.9.9.9 dns.quad9.net
     block list domain https://dbl.oisd.nl/basic/
     allow domain vortex.data.microsoft.com
 }

--- a/action_test.go
+++ b/action_test.go
@@ -1,6 +1,7 @@
 package filter
 
 import (
+	"net/http"
 	"testing"
 )
 
@@ -156,6 +157,7 @@ func TestActionWildcardDNSMasq(t *testing.T) {
 }
 
 func TestListResolver(t *testing.T) {
+	http.DefaultTransport = nil
 	corefile := `filter {
 		listresolver 9.9.9.9
 		block list domain https://dbl.oisd.nl/basic/
@@ -163,6 +165,19 @@ func TestListResolver(t *testing.T) {
 	filter := NewTestFilter(t, corefile)
 	filter.Build()
 	if len(filter.blockDomains) == 0 {
-		t.Errorf("expected no domains; got %d", len(filter.blockDomains))
+		t.Error("expected domains; got none")
+	}
+}
+
+func TestListResolverTLS(t *testing.T) {
+	http.DefaultTransport = nil
+	corefile := `filter {
+		listresolver tls://9.9.9.9 dns.quad9.net
+		block list domain https://dbl.oisd.nl/basic/
+	}`
+	filter := NewTestFilter(t, corefile)
+	filter.Build()
+	if len(filter.blockDomains) == 0 {
+		t.Errorf("expected domains; got none")
 	}
 }

--- a/load.go
+++ b/load.go
@@ -2,6 +2,7 @@ package filter
 
 import (
 	"context"
+	"crypto/tls"
 	"fmt"
 	"io"
 	"net"
@@ -11,6 +12,8 @@ import (
 	"os"
 	"strings"
 	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/transport"
 )
 
 // ListLoader contains the means to retrieve a list
@@ -60,22 +63,37 @@ func (FileListLoader) Load(path string) (io.ReadCloser, error) {
 // HTTPListLoader retrieves lists from remote sources using HTTP/HTTPS
 type HTTPListLoader struct {
 	Network    string
+	ServerName string
 	ResolverIP netip.AddrPort
 }
 
 // Load implements ListLoader
 func (h HTTPListLoader) Load(path string) (io.ReadCloser, error) {
-	client := http.Client{}
+	client := &http.Client{}
 	if h.ResolverIP.IsValid() {
 		dialFunc := func(ctx context.Context, network, address string) (net.Conn, error) {
 			d := net.Dialer{
 				Timeout: time.Second * 5,
 			}
-			return d.DialContext(ctx, h.Network, h.ResolverIP.String())
+			net, err := h.transportToNetwork(h.Network)
+			if err != nil {
+				// This shouldn't be reached. See default return in
+				// HTTPListLoader.transportToNetwork
+				return nil, err
+			}
+			conn, err := d.DialContext(ctx, net, h.ResolverIP.String())
+			if net == "tcp" {
+				tlsConfig := &tls.Config{
+					ServerName: h.ServerName,
+				}
+				return tls.Client(conn, tlsConfig), err
+			}
+			return conn, err
 		}
 		dialerResolver := &net.Resolver{
-			PreferGo: true,
-			Dial:     dialFunc,
+			PreferGo:     true,
+			StrictErrors: true,
+			Dial:         dialFunc,
 		}
 		dialer := &net.Dialer{
 			Resolver: dialerResolver,
@@ -99,4 +117,20 @@ func (h HTTPListLoader) Load(path string) (io.ReadCloser, error) {
 		)
 	}
 	return resp.Body, nil
+}
+
+// convert the dns transport type to the corresponding network used by a dialer.
+// loadTransport is expected to be a constant from
+// github.com/coredns/coredns/plugin/pkg/transport
+func (h HTTPListLoader) transportToNetwork(loadTransport string) (string, error) {
+	switch loadTransport {
+	case transport.DNS:
+		return "udp", nil
+	case transport.TLS:
+		return "tcp", nil
+	default:
+		// This error shouldn't be reached in normal circumstances since the
+		// Corefile parser only sets DNS or TLS transport
+		return "", fmt.Errorf("unknown listresolver transport %q", loadTransport)
+	}
 }

--- a/setup.go
+++ b/setup.go
@@ -11,6 +11,7 @@ import (
 	"github.com/coredns/coredns/core/dnsserver"
 	"github.com/coredns/coredns/plugin"
 	"github.com/coredns/coredns/plugin/pkg/parse"
+	"github.com/coredns/coredns/plugin/pkg/transport"
 	"github.com/miekg/dns"
 )
 
@@ -410,27 +411,33 @@ func parseListResolver(c *caddy.Controller, f *Filter) error {
 	if len(to) == 0 {
 		return c.Errf("no list resolver address specified")
 	}
-	toHosts, err := parse.HostPortOrFile(to...)
+	toHosts, err := parse.HostPortOrFile([]string{to[0]}...)
 	if err != nil {
 		return err
 	}
-	transport, host := parse.Transport(toHosts[0])
+	xprt, host := parse.Transport(toHosts[0])
 	ipaddr, err := netip.ParseAddrPort(host)
 	if err != nil {
 		return err
 	}
-	switch transport {
-	case "dns":
-		f.allowConfig.HTTPLoader.Network = "udp"
-		f.blockConfig.HTTPLoader.Network = "udp"
-	case "tls":
-		f.allowConfig.HTTPLoader.Network = "tcp"
-		f.blockConfig.HTTPLoader.Network = "tcp"
+	switch xprt {
+	case transport.DNS, transport.TLS:
+		f.allowConfig.HTTPLoader.Network = xprt
+		f.blockConfig.HTTPLoader.Network = xprt
 	default:
 		return fmt.Errorf(
 			"%q is not a supported transport for listresolver",
-			transport,
+			xprt,
 		)
+	}
+	if xprt == transport.TLS {
+		if len(to) == 1 {
+			return c.Err(
+				"listresolver is using tls scheme without a server name",
+			)
+		}
+		f.allowConfig.HTTPLoader.ServerName = to[1]
+		f.blockConfig.HTTPLoader.ServerName = to[1]
 	}
 	f.allowConfig.HTTPLoader.ResolverIP = ipaddr
 	f.blockConfig.HTTPLoader.ResolverIP = ipaddr

--- a/setup_test.go
+++ b/setup_test.go
@@ -956,9 +956,16 @@ func TestParserListResolver(t *testing.T) {
 			false,
 		},
 		{
-			"listresolver quad9 tls",
+			"listresolver quad9 tls without server name",
 			`filter {
 				listresolver tls://9.9.9.9
+			}`,
+			true,
+		},
+		{
+			"listresolver quad9 tls without server name",
+			`filter {
+				listresolver tls://9.9.9.9 dns.quad9.net
 			}`,
 			false,
 		},

